subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/LambdaImplementationSnapshot.java

```diff
@@ -20,60 +20,118 @@
 import org.gradle.internal.hash.Hasher;
 
 import javax.annotation.Nullable;
+import java.lang.invoke.SerializedLambda;
+import java.util.Objects;
 
 public class LambdaImplementationSnapshot extends ImplementationSnapshot {
 
-    public LambdaImplementationSnapshot(String typeName) {
-        super(typeName);
+    private final HashCode classLoaderHash;
+
+    private final String functionalInterfaceClass;
+    private final String implClass;
+    private final String implMethodName;
+    private final String implMethodSignature;
+    private final int implMethodKind;
+
+    public LambdaImplementationSnapshot(HashCode classLoaderHash, SerializedLambda lambda) {
+        this(
+            lambda.getCapturingClass(),
+            classLoaderHash,
+            lambda.getFunctionalInterfaceClass(),
+            lambda.getImplClass(),
+            lambda.getImplMethodName(),
+            lambda.getImplMethodSignature(),
+            lambda.getImplMethodKind()
+        );
+    }
+
+    public LambdaImplementationSnapshot(
+        String capturingClass,
+        HashCode classLoaderHash,
+        String functionalInterfaceClass,
+        String implClass,
+        String implMethodName,
+        String implMethodSignature,
+        int implMethodKind
+    ) {
+        super(capturingClass);
+        this.classLoaderHash = classLoaderHash;
+        this.functionalInterfaceClass = functionalInterfaceClass;
+        this.implClass = implClass;
+        this.implMethodName = implMethodName;
+        this.implMethodSignature = implMethodSignature;
+        this.implMethodKind = implMethodKind;
     }
 
     @Override
     public void appendToHasher(Hasher hasher) {
-        throw new RuntimeException("Cannot hash implementation of lambda " + getTypeName());
+        hasher.putString(ImplementationSnapshot.class.getName());
```

Comment:
```suggestion
        hasher.putString(LambdaImplementationSnapshot.class.getName());
```

Similar to `ClassImplementationSnapshot`, we should probably use the concrete class here. Though given that the number of things we hash are different, we are already sure that we don't get a collision.